### PR TITLE
NGFW-13398 2FA/MFA for OpenVPN (RFC 6238)

### PR DESCRIPTION
**Pull request by rasmussen-untangle**
Allow OpenVPN users to create an additional security level by using two factory authentication based on on  RFC 6238

### CODE REVIEW

File: buildtools/jars.rb [hunk 1]
```diff
@@ -97,13 +97,23 @@ def Jars.findJars
                            Jars.downloadTarget('geoip2-2.6.0-with-dependencies/geoip2-2.6.0/lib/maxmind-db-1.2.0.jar'),
                            Jars.downloadTarget('geoip2-2.6.0-with-dependencies/geoip2-2.6.0/lib/geoip2-2.6.0.jar')])
 
+    const_set(:Zxing, [ Jars.downloadTarget('zxing-3.4.1/jcommander-1.78.jar'),
```
Comment (jsommerville-untangle) [inline buildtools/jars.rb:100]: can we use qrencode from wireguard to do this also?

File: openvpn/hier/usr/share/untangle/bin/openvpn-auth-user-pass [hunk 2]
```diff
@@ -25,7 +26,16 @@ if (username == None):
...
 if (password == None):
     raise Exception("Missing password")
-    
-result = openvpnApp.userAuthenticate(username,password)
+
+# Figure out the OTP password.
+# OpenVPN clients will combine password and OTP into the password string.
+# Format is "SCRV1:<passwd_b64>:<otp_b64>"
+otp = 0
+if password.startswith("SCRV1:"):
```
Comment (jsommerville-untangle) [inline openvpn/hier/usr/share/untangle/bin/openvpn-auth-user-pass:34]: maybe check for parts splitting as well as startswith, is the password encoded at this point (for non OTP users)?

File: openvpn/js/view/Server.js [hunk 3]
```diff
@@ -91,6 +91,15 @@ Ext.define('Ung.apps.openvpn.view.Server', {
                 value: '{settings.authUserPass}',
                 disabled: '{settings.serverEnabled == false}'
             },
+        },            {
+            xtype: 'checkbox',
+            boxLabel: 'Add TOTP prompt keyword to client config',
```
Comment (jsommerville-untangle) [inline openvpn/js/view/Server.js:96]: is there a better user identifiable string for this label? "Add TOTP Configuration" or something?

File: openvpn/src/com/untangle/app/openvpn/OpenVpnSettings.java [hunk 4]
```diff
@@ -43,6 +43,7 @@
     private boolean serverEnabled = false;
     private boolean natOpenVpnTraffic = true;
     private boolean authUserPass = false;
+    private boolean toptClientPrompt = false;
```
Comment (jsommerville-untangle) [inline openvpn/src/com/untangle/app/openvpn/OpenVpnSettings.java:46]: this variable is "toptClient" should maybe be consistent with "TOTP" usage

File: uvm/impl/com/untangle/uvm/LocalDirectoryImpl.java [hunk 5]
```diff
@@ -159,6 +186,83 @@ public String testRadiusProxyLogin(String userName, String userPass, String user
...
+     * @param username
+     *        The username used in the new key.
+     * @param issuer
+     *        The issuer or hostname.
+     * @param secret
+     *        The secret in Base32 encoding.
+     * 
+     * @return URL string.
+     */
+    public String showSecretQR(String username, String issuer, String secret) {
```
Comment (jsommerville-untangle) [inline uvm/impl/com/untangle/uvm/LocalDirectoryImpl.java:215]: see if we can reuse wiregard QR code generation

File: uvm/servlets/admin/config/local-directory/MainController.js [hunk 6]
```diff
@@ -212,6 +212,11 @@ Ext.define('Ung.config.local-directory.MainController', {
         });
     },
 
+    generateKey: function(cmp) {
```
Comment (jsommerville-untangle) [inline uvm/servlets/admin/config/local-directory/MainController.js:215]: remove

File: uvm/servlets/admin/config/local-directory/view/Users.js [hunk 7]
```diff
@@ -86,6 +87,16 @@ Ext.define('Ung.config.local-directory.view.Users', {
                 return result;
             },this)
         }, {
+            header: 'Two factor secret'.t(),
+            width: Renderer.messageWidth,
+            dataIndex: 'twofactorSecretKey',
+            renderer: Ext.bind(function(value, metadata, record) {
```
Comment (jsommerville-untangle) [inline uvm/servlets/admin/config/local-directory/view/Users.js:93]: can this be a generic render used in password and two factor secret?

**Review by jsommerville-untangle — APPROVED** [review]

File: openvpn/hier/usr/share/untangle/bin/openvpn-auth-user-pass (hunk 2 above)
Comment (rasmussen-untangle) [reply openvpn/hier/usr/share/untangle/bin/openvpn-auth-user-pass:34]: The password is encoded at this point, so this check works, even if a user has a password that starts with "SCRV1:"

File: openvpn/js/view/Server.js [hunk 8]
```diff
@@ -91,6 +91,21 @@ Ext.define('Ung.apps.openvpn.view.Server', {
                 value: '{settings.authUserPass}',
                 disabled: '{settings.serverEnabled == false}'
             },
+        },{
+            xtype: 'checkbox',
+            boxLabel: 'Add MFA client configuration',
```
Comment (cblaise-untangle) [inline openvpn/js/view/Server.js:96]: Add this to translation with 'sting'.t()
(check all other user facing strings)

File: uvm/servlets/admin/config/local-directory/view/Users.js [hunk 9]
```diff
@@ -78,13 +79,41 @@ Ext.define('Ung.config.local-directory.view.Users', {
...
+            }, this)
+        },
+        {
+            header: 'OpenVPN MFA secret'.t(),
+            xtype: 'actioncolumn',
+            width: Renderer.messageWidth,
+            dataIndex: 'twofactorSecretKey',
+            iconCls: 'fa fa-cog',
+            align: 'center',
+            tooltip: 'Enable/disable MFA by editting user.',
```
Comment (cblaise-untangle) [inline uvm/servlets/admin/config/local-directory/view/Users.js:97]: translate

File: uvm/servlets/admin/config/local-directory/view/Users.js [hunk 10]
```diff
@@ -164,6 +193,57 @@ Ext.define('Ung.config.local-directory.view.Users', {
...
+                width: 500,
+                allowBlank: true,
+                emptyText: '[Key not currently set]'.t(),
+                bind: {
+                    value: '{record.twofactorSecretKey}',
+                    hidden: '{!record.username || !record.mfaEnabled}'
+                },
+            }, {
+                xtype: 'button',
+                text: 'Generate new key',
```
Comment (cblaise-untangle) [inline uvm/servlets/admin/config/local-directory/view/Users.js:221]: translate

**Review by cblaise-untangle — APPROVED** [review]

**Pull request merged**